sdks/python/apache_beam/runners/worker/opcounters.py

```diff
@@ -24,11 +24,99 @@
 import math
 import random
 
+from apache_beam.utils import counters
 from apache_beam.utils.counters import Counter
+from apache_beam.utils.counters import CounterName
 
 # This module is experimental. No backwards-compatibility guarantees.
 
 
+class TransformIoCounter(object):
+  """Class to track time and bytes consumed while reading from IO.
+
+  Examples of IO can be side inputs, shuffle, or streaming state.
+  """
+
+  def check_step(self):
+    pass
+
+  def __exit__(self, exc_type, exc_value, traceback):
+    pass
+
+  def __enter__(self):
+    pass
+
+  def add_bytes_read(self, n):
+    pass
+
+
+class SideInputReadCounter(TransformIoCounter):
+  """Tracks time and bytes consumed while reading from side inputs.
+
+  This class is designed to track consumption of side inputs across fused steps.
+  We represent a side input as a declaring step, and an input index, which are
+  represented as an io_target.
+  The declaring step is the step that originally receives the side input for
+  consumption, and the input index in which the declaring step receives the side
+  input that we want to identify.
+  Note that the declaring step originally receives the side input, but it may
+  not be the only step that spends time reading from this side input.
+  """
+
+  def __init__(self, counter_factory, state_sampler, declaring_step,
+               input_index):
+    """Create a side input read counter.
+
+    Args:
+      counter_factory: A counters.CounterFactory to create byte counters.
+      state_sampler: A statesampler.StateSampler to transition into read states.
+      declaring_step: The step that directly receives the side input initially.
+      input_index: The index of the side input in the list of inputs of the
+        declaring step.
```

Comment:
<!--thread_id:cc_158386095_t; commit:1e64e1f8b880a9fe4450c84cf1c7c5694eb3567e; resolved:0-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> Is this a stable thing to refer to, i.e. is the order of the side inputs guaranteed?  I recall that side inputs are keyed by a string label / key in `si_tags_and_types` (see https://github.com/apache/beam/blob/1882939c0c240d8d1fff654f3f46be59b87ba1d2/sdks/python/apache_beam/runners/dataflow/dataflow_runner.py#L608).

<!--section:body-->
Yes. The order of the side inputs is guaranteed. 

In the design phase of this project, we decided to use the (declaring_step, input index) pair to identify a side input because it's a concept that fits well within DFE. PCollections only have a Tag associated in Dataflow (not in Flume). We can get on GVC and have a chat about the design considerations if you'd like.